draft-irtf-cfrg-vdaf.md

```diff
@@ -1190,6 +1190,198 @@ how the nonces are chosen, but security requires that the nonces be unique. See
+"Helper". The high-level idea is that the Leader and Helper will take turns
+running the computation locally until input from their peer is required:
+
+* For a 1-round VDAF (e.g., Prio3 ({{prio3}})), the Leader sends its prep share
+  to the Helper, who computes the prep message locally (the computations are
+  specified in the following sections), computes its output share, then sends
+  the prep message to the Leader. Preparation requires just one request from
+  the Leader to the Helper.
+
+* For a 2-round VDAF (e.g., Poplar1 ({{poplar1}})), the Leader sends its
+  first-round prep share to the Helper, who replies with the first-round prep
+  message and its second-round prep share. In the next request, the Leader
+  computes its second-round prep share locally, computes its output share, and
+  sends the second-round prep message to the Helper. Finally, the Helper
+  computes its own output share. Preparation requires two requests.
+
+* In general, each request includes the Leader's prep share for the previous
+  round and/or the prep message for the current round; correspondingly, each
+  response consists of the prep message for the current round and the Helper's
+  prep share for the next round.
+
+The Aggregators proceed in this ping-ponging fashion until a step of the
+computation fails, indicating the report is invalid and should be rejected, or
+the last round is reached and each Aggregator computes an output share. All
+told there there are `ceil((Vdaf.ROUNDS+1)/2)` requests sent.
+
+The message types are as follows:
+
+* `PingPongMessage(prep_share)` - Sent by the Leader to the Helper in the
+  initial request. The payload includes its first-round prep share.
+
+* `PingPongContinue(prep_msg, prep_share)` - Sent by the Leader or Helper in an
+  intermediate request or response. The payload includes the prep message from
+  the current round and the sender's prep share for the next round.
+
+* `PingPongFinish(prep_msg)` - Sent by the Leader in the last request
+  (even-round VDAF) or the Helper in the last response (odd-round VDAF). The
+  payload is the final-round prep message.
+
+* `PingPongFinished()` - Sent by the Helper in the final response in even-round
+  VDAFs. There is no payload.
+
+Each Aggregator handles these messages according to the following transition rule:
+
+~~~ transition
+def ping_pong_next(Vdaf,
+                   is_leader: bool,
+                   agg_param: Vdaf.AggParam,
+                   state: State,
+                   inbound: Optional[Message],
+                   ) -> (State, Optional[Message]):
+    if type(inbound) == PingPongMessage:
+        prep_msg, peer_prep_share = None, inbound.prep_share
+    elif type(inbound) == PingPongContinue:
+        prep_msg = inbound.prep_msg
+        peer_prep_share = inbound.prep_share
+    elif type(inbound) == PingPongFinish:
+        prep_msg, peer_prep_share = inbound.prep_msg, None
+    else: # inbound == None, type(inbound) == PingPongFinished
```

Comment:
I think `type(inbound) == PingPongFinished` here should be unreachable when executing this between two honest aggregators. (setting that aside, calling prep_next in such a situation would be incorrect) Only the helper sends PingPongFinished, thus only the leader can receive it. Walking backwards in the protocol, the leader would have transitioned to the Finished state before sending PingPongFinish for the prior round, so the leader would no longer be in the Continued state. Thus, if the leader receives a PingPongFinished message, it will not be calling `ping_pong_next()` with it as the inbound message.

We should have a separate branch here that transitions to Rejected and returns if either aggregator receives PingPongFinished while they're still in the Continued state.

Reply:
Done. I added the rejection branch, but did not condition on being in the Continued state, since we don't expect ping_pong_next() to be called in any ogther state.

Reply:
Oh wait, I'm now realizing htat's wrong; the Leder will be in Finished() while awaiting PingPongFinished().

Reply:
On third thought, currently we say that after reaching Finished the Leader won't await any other messages. I think this fine, but we might consider having the Leader go to Continued while waiting for PingPongFinished. WDYT?

Reply:
I think we can have the Leader go to Finished; going to Continued is only tricky because it stores a prep state instead of an out share (so we'd really have two states, a "normal continued" state and a "continued awaiting finished" state, I suppose).

Maybe we should just say that once the Leader goes to the Finished state, it should expect one more PingPongFinished message? I think that is effectively the semantics we want.

Reply:
Having the leader go to Finished makes sense to me -- otherwise we'd need two separate state machines for the leader and helper. The helper only sends PingPongFinished because we have to send something in a request-response transport. Since the leader should already be in a terminal state machine state, the message can be dropped on the floor -- the VDAF is already done as far as all parties are concerned.

Reply:
SGTM. Added a note about even-round VDAFs.